Add schema to the current form so that they will work for Plone 5

prakharjoshi opened the pull request:
Adding Mail schema to forms tried it on Plone 5 instance and it worked for me on local. Have a look at this.
@fulv

fulv commented:
Thanks @prakharjoshi, I will look at it later today.

prakharjoshi commented:
@fulv  Please have a look at the travis logs, I guess it is related to some other packages dependencies, I think this add-on needs to be updated, what you say ?

fulv commented:
It does not work for me, it keeps bringing up the general Plone mail control panel, and this:

```
2016-04-01 08:51:32 ERROR Zope.SiteErrorLog 1459500692.080.271574954811 http://localhost:8080/Plone2/++rule++rule-1/++action++0/edit
Traceback (innermost last):
  Module ZPublisher.Publish, line 138, in publish
  Module ZPublisher.mapply, line 77, in mapply
  Module Products.PDBDebugMode.runcall, line 70, in pdb_runcall
  Module ZPublisher.Publish, line 48, in call_object
  Module z3c.form.form, line 233, in __call__
  Module plone.z3cform.fieldsets.extensible, line 59, in update
  Module plone.z3cform.patch, line 30, in GroupForm_update
  Module z3c.form.group, line 132, in update
  Module z3c.form.form, line 136, in updateWidgets
  Module z3c.form.field, line 243, in update
  Module z3c.form.datamanager, line 102, in canWrite
  Module z3c.form.datamanager, line 66, in adapted_context
TypeError: ('Could not adapt', <MailGroupAction at /Plone2/++rule++rule-1/++action++0>, <InterfaceClass Products.CMFPlone.interfaces.controlpanel.IMailSchema>)
> /home/vagrant/var/buildout-cache/eggs/z3c.form-3.2.9-py2.7.egg/z3c/form/datamanager.py(66)adapted_context()
-> context = self.field.interface(context)
```

I think it's because of this:

```
    schema = IMailSchema
```

fulv commented:
It does not work for me, it keeps bringing up the general Plone mail control panel, and this:

```
2016-04-01 08:51:32 ERROR Zope.SiteErrorLog 1459500692.080.271574954811 http://localhost:8080/Plone2/++rule++rule-1/++action++0/edit
Traceback (innermost last):
  Module ZPublisher.Publish, line 138, in publish
  Module ZPublisher.mapply, line 77, in mapply
  Module Products.PDBDebugMode.runcall, line 70, in pdb_runcall
  Module ZPublisher.Publish, line 48, in call_object
  Module z3c.form.form, line 233, in __call__
  Module plone.z3cform.fieldsets.extensible, line 59, in update
  Module plone.z3cform.patch, line 30, in GroupForm_update
  Module z3c.form.group, line 132, in update
  Module z3c.form.form, line 136, in updateWidgets
  Module z3c.form.field, line 243, in update
  Module z3c.form.datamanager, line 102, in canWrite
  Module z3c.form.datamanager, line 66, in adapted_context
TypeError: ('Could not adapt', <MailGroupAction at /Plone2/++rule++rule-1/++action++0>, <InterfaceClass Products.CMFPlone.interfaces.controlpanel.IMailSchema>)
> /home/vagrant/var/buildout-cache/eggs/z3c.form-3.2.9-py2.7.egg/z3c/form/datamanager.py(66)adapted_context()
-> context = self.field.interface(context)
```

I think it's because of this:

```
    schema = IMailSchema
```

fulv commented:
It does not work for me, it keeps bringing up the general Plone mail control panel, and this:

```
2016-04-01 08:51:32 ERROR Zope.SiteErrorLog 1459500692.080.271574954811 http://localhost:8080/Plone2/++rule++rule-1/++action++0/edit
Traceback (innermost last):
  Module ZPublisher.Publish, line 138, in publish
  Module ZPublisher.mapply, line 77, in mapply
  Module Products.PDBDebugMode.runcall, line 70, in pdb_runcall
  Module ZPublisher.Publish, line 48, in call_object
  Module z3c.form.form, line 233, in __call__
  Module plone.z3cform.fieldsets.extensible, line 59, in update
  Module plone.z3cform.patch, line 30, in GroupForm_update
  Module z3c.form.group, line 132, in update
  Module z3c.form.form, line 136, in updateWidgets
  Module z3c.form.field, line 243, in update
  Module z3c.form.datamanager, line 102, in canWrite
  Module z3c.form.datamanager, line 66, in adapted_context
TypeError: ('Could not adapt', <MailGroupAction at /Plone2/++rule++rule-1/++action++0>, <InterfaceClass Products.CMFPlone.interfaces.controlpanel.IMailSchema>)
> /home/vagrant/var/buildout-cache/eggs/z3c.form-3.2.9-py2.7.egg/z3c/form/datamanager.py(66)adapted_context()
-> context = self.field.interface(context)
```

I think it's because of this:

```
    schema = IMailSchema
```

prakharjoshi commented:
Hi Fulvio,

I added the mail Schema (IMailSchema) and it was working for me. I added it
here
https://github.com/collective/collective.contentrules.mailtogroup/pull/9/files

Cheers,
Prakhar Joshi
DA-IICT,Gandhinagar

On Fri, Apr 1, 2016 at 2:46 PM, Fulvio Casali <notifications@github.com>
wrote:

> It does not work for me, it keeps bringing up the general Plone mail
> control panel, and this:
>
> 2016-04-01 08:51:32 ERROR Zope.SiteErrorLog 1459500692.080.271574954811 http://localhost:8080/Plone2/++rule++rule-1/++action++0/edit
> Traceback (innermost last):
>   Module ZPublisher.Publish, line 138, in publish
>   Module ZPublisher.mapply, line 77, in mapply
>   Module Products.PDBDebugMode.runcall, line 70, in pdb_runcall
>   Module ZPublisher.Publish, line 48, in call_object
>   Module z3c.form.form, line 233, in __call__
>   Module plone.z3cform.fieldsets.extensible, line 59, in update
>   Module plone.z3cform.patch, line 30, in GroupForm_update
>   Module z3c.form.group, line 132, in update
>   Module z3c.form.form, line 136, in updateWidgets
>   Module z3c.form.field, line 243, in update
>   Module z3c.form.datamanager, line 102, in canWrite
>   Module z3c.form.datamanager, line 66, in adapted_context
> TypeError: ('Could not adapt', <MailGroupAction at /Plone2/++rule++rule-1/++action++0>, <InterfaceClass Products.CMFPlone.interfaces.controlpanel.IMailSchema>)
> > /home/vagrant/var/buildout-cache/eggs/z3c.form-3.2.9-py2.7.egg/z3c/form/datamanager.py(66)adapted_context()
> -> context = self.field.interface(context)
>
> I think it's because of this:
>
>     schema = IMailSchema
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/collective/collective.contentrules.mailtogroup/pull/9#issuecomment-204323575>
>

fulv commented:
I believe this is ready for a 2.0 release for Plone 5, and also a 1.5.1 release for Plone 4.
Anyone want to code-review / merge / release, or add me as an owner in pypi so I can make the release?

@mauritsvanrees @kcleong @hvelarde @cekk @prakharjoshi

prakharjoshi commented:
Tried. Looks good. :)

mauritsvanrees commented:
I have not tested it, but it seems okay. And a 1.5.x branch would be good, without these changes.

PyPI owners are: @puittenbroek @khink @huubbouma @helmantel @kcleong

khink commented:
@fulv what's your pypi id?

fulv commented:
It's fulv, as well.
Thanks!

On Mon, Apr 4, 2016 at 1:41 PM Kees Hink <notifications@github.com> wrote:

> @fulv <https://github.com/fulv> what's your pypi id?
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/collective/collective.contentrules.mailtogroup/pull/9#issuecomment-205259800>
>

hvelarde commented on .travis.yml:
```diff
@@ -1,17 +1,12 @@
 language: python
 python: 2.7
 env:
-  - PLONE_VERSION=4.2
```
you're only testing effectively for Plone 5.0; are you dropping compatibility with previous versions?

hvelarde commented on CHANGES.rst:
```diff
@@ -1,6 +1,25 @@
 Changelog
 =========
 
+2.0 (unreleased)
```
if the above is true, I don't see that reflected here in the changelog

hvelarde commented on buildout.cfg:
```diff
@@ -1,33 +1,8 @@
-    https://raw.github.com/collective/buildout.plonetest/master/test-4.x.cfg
-    https://raw.github.com/collective/buildout.plonetest/master/qa.cfg
-
+extends = https://raw.githubusercontent.com/collective/buildout.plonetest/master/test-5.x.cfg
```
you broke the testing of 4.3 branch with these changes

mauritsvanrees replied:
I wondered about the changes in theTravis file too, but the 4.3 tests pass: https://travis-ci.org/collective/collective.contentrules.mailtogroup/jobs/120451498

mauritsvanrees replied:
Ah, I see: the 4.3 tests on Travis pass, but that is only because they actually use Plone 5...
The `bin/buildout annotate` line shows that Products.CMFPlone 5.0.3.1 is being used.

fulv replied:
Can you guys please explain this to me?  I have no idea how this travis stuff works.  Why do 4.3 tests use Plone 5?

hvelarde commented on collective/contentrules/mailtogroup/actions/mail.py:
```diff
@@ -96,16 +103,75 @@ def __init__(self, context, element, event):
+        try:
+            mailhost.send(mime_msg)
+
+        # Plone-3
```
get rid of this, please: Plone 3 is not supported anymore.

fulv replied:
Gladly.

hvelarde commented:
@fulv @prakharjoshi in my experience it's always better not to mix QA stuff with features/bug fixes as code review is harder.

I would suggest to split this PR into two different: one to achieve Plone 5 compatibility and one for QA stuff.

prakharjoshi commented on collective/contentrules/mailtogroup/actions/mail.py:
```diff
@@ -96,16 +103,75 @@ def __init__(self, context, element, event):
         self.context = context
```
@fulv  here at line [100](https://github.com/collective/collective.contentrules.mailtogroup/pull/9/files#diff-4c9070de35a2ba0217cedb5665f50917R100) There are two inheritance of Interface class, we can remove one, right ?

fulv replied:
I believe it's needed, because the MailActionExecutor is a multi-adapter from a context, an action and an event.

khink commented:
@fulv you're an Owner on PyPi now.

fulv commented:
@hvelarde But if I split into two PRs, then the first one's tests will always fail.  Also, I don't see the benefit, since the changes are cleanly spearated by files (mail.py and test_action_mail_to_group.py).

hvelarde commented:
@fulv please review and fix #10; it's almost done and, you don't have to fix Flake8 and all other QA stuff now. keeping compatibility with Plone 4 and 5 in a content rule is very easy, so I'll be against maintaining 2 different branches.

that's my recommendation to keep things simpler in the future but, as this is not my package, feel free to do whatever you decide. I just fixed CI to let you be aware of what you're doing.

prakharjoshi commented:
@hvelarde  I guess this improvement will work till plone 4.3 and plone 5 too. Do we need backward compatibility with plone < 4.3 too ? Then we need to put the try/catch to import the AddForm and AddEditForm in mail.py .

hvelarde commented:
@prakharjoshi no, I removed those to make it easier to maintain the package.

fulv commented:
#10 took care of this.
